File: music-application/src/main/java/com/music/sale/application/product/service/ProductCommandService.java
package com.music.sale.application.product.service;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.music.sale.application.product.dto.input.CreateProductInput;
import com.music.sale.application.product.dto.input.UpdateProductInput;
import com.music.sale.application.product.dto.output.ProductOutput;
import com.music.sale.application.product.mapper.ProductMapper;
import com.music.sale.application.product.port.inport.ProductCommandUseCase;
import com.music.sale.application.product.port.outport.ProductCommandPort;
import com.music.sale.application.product.port.outport.ProductQueryPort;
import com.music.sale.domain.product.ProductItem;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
@Transactional
public class ProductCommandService implements ProductCommandUseCase {
    private final ProductCommandPort commandPort;
    private final ProductQueryPort queryPort;
    private final ProductMapper mapper;

    @Override
    public ProductOutput createProduct(CreateProductInput input, Long currentUserId) {
        ProductItem item = mapper.toDomainForCreate(input, currentUserId);
        ProductItem saved = commandPort.saveProduct(item);
        return mapper.toOutput(saved);
    }

    @Override
    public ProductOutput updateProduct(Long productId, UpdateProductInput input, Long currentUserId) {
        ProductItem existing = queryPort.findByProductId(productId)
                .orElseThrow(() -> new IllegalArgumentException("상품을 찾을 수 없습니다. productId=" + productId));

        if (!existing.getSellerId().equals(currentUserId)) {
            throw new IllegalArgumentException("수정 권한이 없습니다.");
        }

        ProductItem updated = mapper.toDomainForUpdate(existing, input);
        ProductItem saved = commandPort.updateProduct(updated);
        return mapper.toOutput(saved);
    }

    @Override
    public void deleteProduct(Long productId, Long currentUserId) {
        ProductItem existing = queryPort.findByProductId(productId)
                .orElseThrow(() -> new IllegalArgumentException("상품을 찾을 수 없습니다. productId=" + productId));

        if (!existing.getSellerId().equals(currentUserId)) {
            throw new IllegalArgumentException("삭제 권한이 없습니다.");
        }

        commandPort.deleteByProductId(productId);
    }
}

File: music-application/src/main/kotlin/com/music/sale/application/product/service/ProductCommandService.kt
package com.music.sale.application.product.service

import com.music.sale.application.category.service.CategoryService
import com.music.sale.application.product.dto.CreateProductInput
import com.music.sale.application.product.dto.ProductOutput
import com.music.sale.application.product.dto.UpdateProductInput
import com.music.sale.application.product.mapper.ProductMapper
import com.music.sale.application.product.port.inport.ProductCommandUseCase
import com.music.sale.application.product.port.outport.ProductCommandPort
import com.music.sale.application.product.port.outport.ProductQueryPort
import com.music.sale.application.user.port.outport.UserPort
import com.music.sale.domain.store.Store
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
open class ProductCommandService(
    private val commandPort: ProductCommandPort,
    private val queryPort: ProductQueryPort,
    private val mapper: ProductMapper,
    private val categoryService: CategoryService,
    private val userPort: UserPort,
) : ProductCommandUseCase {
    override fun createProduct(input: CreateProductInput): ProductOutput {
        val seller =
            userPort.findById(input.sellerId)
                ?: throw IllegalArgumentException(
                    "Seller not found with id: ${input.sellerId}",
                )

        // TODO: Store 정보도 실제로 조회해야 함
        val store = Store(input.storeId)

        val saveCondition = mapper.toSaveProductCondition(input, seller, store)
        val savedProduct = commandPort.save(saveCondition)

        return mapper.toOutput(savedProduct)
    }

    override fun updateProduct(input: UpdateProductInput): ProductOutput {
        val product = queryPort.findById(input.id) ?: throw IllegalArgumentException("Product not found")

        val updatedProduct =
            product.copy(
                price = input.price ?: product.price,
                condition = input.condition ?: product.condition,
                conditionGrade = input.conditionGrade ?: product.conditionGrade,
                stockQuantity = input.stockQuantity ?: product.stockQuantity,
                status = input.status ?: product.status,
                customName = input.name ?: product.customName,
                customAttributes = input.attributes ?: product.customAttributes,
            )

        return mapper.toOutput(commandPort.update(updatedProduct))
    }

    override fun deleteProduct(id: Long): ProductOutput {
        val product = queryPort.findById(id) ?: throw IllegalArgumentException("Product not found")
        commandPort.deleteById(id)
        return mapper.toOutput(product)
    }
} 
